File: pe/nonvisualstudio/clojure/Implementation/Problem0006.clj
(ns pe.clojure.implementation.problem0006)

(load-file "C:/Users/Thomas/Documents/GitHub/practice/utilities/nonvisualstudio/clojure/Implementation/Sequences.clj")
(use '[utilities.sequences :as util.seq])

(defn square [x]
  (* x x))

(defn sumOfSquares [xInts]
  (->>
   (util.seq/firstInts xInts)
   (map square)
   util.seq/sum))

(defn squareOfSums [xInts]
  (->>
   (util.seq/firstInts xInts)
   util.seq/sum
   square))

(defn differenceBetweenSquareOfSumsAndSumOfSquares [xInts]
  (-
   (squareOfSums xInts)
   (sumOfSquares xInts)))

File: pe/nonvisualstudio/elixir/Implementation/Problem0006.ex
defmodule Problem0006 do
  def sumOfSquaresOfNumberLine(lastNumber) do
    1..lastNumber
    |> Enum.map( &(square(&1)))
    |> Enum.sum
  end

  def squareOfSumsOfNumberLine(lastNumber) do
    1..lastNumber
    |> Enum.sum
    |> square
  end

  def differenceBetweenSquareOfSumsAndSumOfSquares(lastNumber) do
    squareOfSumsOfNumberLine(lastNumber) - sumOfSquaresOfNumberLine(lastNumber)
  end

  defp square(number), do: number * number
end
